tests/phpunit/tests/functions.php

```diff
@@ -558,9 +558,57 @@ public function test_add_query_arg() {
 			);
 		}
 
+		$_SERVER['REQUEST_URI'] = 'nothing';
+
+		$this->assertSame( 'baz=1&foo=1', add_query_arg( 'foo', '1', 'baz=1' ) ); // #WP4903
+
 		$_SERVER['REQUEST_URI'] = $old_req_uri;
 	}
 
+	/**
+	 * @ticket 58902
+	 * @group add_query_arg
+	 */
+	public function test_add_query_arg_sanitize_url() {
+		$old_request_uri = $_SERVER['REQUEST_URI'];
+
+		$urls_with_query_string = array(
+			'http://example.com/two words?foo=1' => 'http://example.com/two%20words?foo=1',
+			'http;//example.com?foo=1'           => 'http://example.com?foo=1',
+			'example.com?foo=1'                  => 'http://example.com?foo=1',
+			'/example?foo=1'                     => '/example?foo=1',
+			' ?foo=1'                            => '?foo=1',
+			' example.com?foo=1'                 => 'http://example.com?foo=1',
+			'http://example.com?foo=[1]'         => 'http://example.com?foo=%5B1%5D',
+		);
+
+		foreach ( $urls_with_query_string as $wrong_url => $correct_url ) {
+			$_SERVER['REQUEST_URI'] = $wrong_url;
+			$this->assertSame( "$correct_url&bar=2", add_query_arg( array( 'bar' => '2' ) ) );
+		}
+
+		$urls_without_query_string = array(
+			'/'                              => '/',
+			'example.com'                    => 'http://example.com',
+			' example.com'                   => 'http://example.com',
+			'http;//example.com'             => 'http://example.com',
+			'http://example.com/two words'   => 'http://example.com/two%20words',
+			'wrongprotocol://example.com'    => '',
+			'/example'                       => '/example',
+			'example.php'                    => 'example.php',
+			'example'                        => 'http://example',
+			'http://example/%0d%0a%0D%0A%20' => 'http://example/%20',
+			'mailto:%0d%0a%0D%0A%20'         => 'mailto:%0d%0a%0D%0A%20',
+		);
+
+		foreach ( $urls_without_query_string as $wrong_url => $correct_url ) {
+			$_SERVER['REQUEST_URI'] = $wrong_url;
+			$this->assertSame( "$correct_url?foo=1", add_query_arg( array( 'foo' => '1' ) ) );
+		}
+
+		$_SERVER['REQUEST_URI'] = $old_request_uri;
+	}
```

Comment:
A few notes:
- The `REQUEST_URI` is only restored after the assertions. That means if any of the assertions failed, `REQUEST_URI` will not be restored.
  1. Store the return value of `add_query_arg()` in an `$actual` variable.
  2. Restore `REQUEST_URI`.
  3. Perform the assertion using `$actual` as the second argument.
- Rather than `@group add_query_arg`, we should be using `@covers ::add_query_arg` instead.
- This test would benefit from using a data provider.
  - Currently, if one assertion fails, the rest will not run. This could make it difficult to get a full picture of any regressions.

For example, place this below the test method:
```php
/**
 * Data provider.
 *
 * @return array[]
 */
public function data_add_query_arg_sanitize_url() {
	return array(
		// URLs with query string.
		'URL with a space in the path'  => array(
			'url'      => 'http://example.com/two words?foo=1',
			'expected' => 'http://example.com/two%20words?foo=1',
		),
		'URL with a semi-colon in the scheme'         => array(
			'url'      => 'http;//example.com?foo=1',
			'expected' => 'http://example.com?foo=1',
		),
		'URL with no scheme'            => array(
			'url'      => 'example.com?foo=1',
			'expected' => 'http://example.com?foo=1',
		),
		'a relative URL'                => array(
			'url'      => '/example?foo=1',
			'expected' => '/example?foo=1',
		),
		'a space and query params'      => array(
			'url'      => ' ?foo=1',
			'expected' => '?foo=1',
		),
		'URL with space before the domain and no scheme' => array(
			'url'      => ' example.com?foo=1',
			'expected' => 'http://example.com?foo=1',
		),
		'URL with an array query param' => array(
			'url'      => 'http://example.com?foo=[1]',
			'expected' => 'http://example.com?foo=%5B1%5D',
		),

		// URLs without query string.
		'just a forward slash' => array(
			'url'      => '/',
			'expected' => '/',
		),
		'URL without query string, no scheme' => array(
			'url'      => 'example.com',
			'expected' => 'http://example.com',
		),
		'URL without query string, no scheme and a leading space' => array(
			'url'      => ' example.com',
			'expected' => 'http://example.com',
		),
		'URL without query string, a semi-colon in the scheme' => array(
			'url'      => 'http;//example.com',
			'expected' => 'http://example.com',
		),
		'URL without query string, a space in the path' => array(
			'url'      => 'http://example.com/two words',
			'expected' => 'http://example.com/two%20words',
		),
		'URL without query string, an invalid scheme' => array(
			'url'      => 'wrongprotocol://example.com',
			'expected' => '',
		),
		'a relative URL without query string' => array(
			'url'      => '/example',
			'expected' => '/example',
		),
		'a single file without query string' => array(
			'url'      => 'example.php',
			'expected' => 'example.php',
		),
		'URL without query string, no scheme or TLD' => array(
			'url'      => 'example',
			'expected' => 'http://example',
		),
		'URL without query string, and an encoded path' => array(
			'url'      => 'http://example/%0d%0a%0D%0A%20',
			'expected' => 'http://example/%20',
		),
		'Mail link without query string, and encoded characters' => array(
			'url'      => 'mailto:%0d%0a%0D%0A%20',
			'expected' => 'mailto:%0d%0a%0D%0A%20',
		),
	);
}
```

The test would then become:
```php
/**
 * @ticket 58902
 *
 * @covers ::add_query_arg
 *
 * @dataProvider data_add_query_arg_sanitize_url
 *
 * @param string $url      The URL.
 * @param string $expected The expected URL.
 */
public function test_add_query_arg_sanitize_url( $url, $expected ) {
	$old_request_uri        = $_SERVER['REQUEST_URI'];
	$_SERVER['REQUEST_URI'] = $url;
	$actual                 = add_query_arg( array( 'bar' => '2' ) );
	$_SERVER['REQUEST_URI'] = $old_request_uri;

	$expected = str_contains( $expected, '?' ) ? "$expected&bar=2" : "$expected?bar=2";
	$this->assertSame( $expected, $actual );
}
```